### DISCUSSION TITLE
"is empty" pattern keyword

### DISCUSSION THREAD

**Discussion by simon-curtis** [dotnet/csharplang#6137] [Language Ideas]
I think it would be nice to have a "is empty" pattern as syntactic sugar, basically a proxy for `{ Length: 0 }`  or `{ Count: 0 }`. What do you people think?

```c#

static bool IsStringNullOrEmpty(string? input)
{
    return input is null or empty;
}

IsStringNullOrEmpty(null)     // true
IsStringNullOrEmpty("")        // true

static bool IsArrayNullOrEmpty<T>(T[]? input)
{
    return input is null or empty;
}

IsArrayNullOrEmpty<string>((string[]?)null)     // true
IsArrayNullOrEmpty<string>(new string[0])     // true
```

Or is this going to be covered by the [C# 11 List Patterns](https://devblogs.microsoft.com/dotnet/early-peek-at-csharp-11-features/#c-11-preview-list-patterns)

like this:

```c#

static bool IsStringNullOrEmpty(string? input)
{
    return input is null or [ ];
}

```

**Comment by jnm2** [comment]
I've been using `is null or []` and I like it. It's intuitive.

**Reply by HaloFour** [reply]
Makes me think that we need to nail down the order of evaluation of subpatterns.  It's seemingly innocuous in this case but this is definitely another situation where one expects the evaluation to be left-to-right.

**Comment by simon-curtis** [comment]
An extension of this of course is that if it is empty then it's implicitally not null, no need for the null check first 

```c#

static bool IsNotNullAndEmpty(string? input)
{
    return input is empty;
    // rather than: return input?.Length is 0;
}
```

**Reply by jnm2** [reply]
How often do you check for something being empty and not null?

**Reply by simon-curtis** [reply]
Rarely, if ever. Just a happy coincidence if someone wanted it

**Comment by 333fred** [comment]
This is where active patterns will come in, imo. I don't think there's a good reason to add very specific patterns like this to the language. In general, after list patterns, I see very few things left to do in the space: expand list patterns to IEnumerable, dictionary patterns, and finally active patterns. After that, I think new types of patterns are pretty much unnecessary.

**Reply by simon-curtis** [reply]
Yeah, as long as string can be implictly converted to char[] then `return input is null or [ ]` would be fine.

Also what are "active patterns" I haven't heard the term

**Reply by 333fred** [reply]
https://github.com/dotnet/csharplang/issues/4131
https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/active-patterns
